ledger-core/insolar/payload/validation.go

```diff
@@ -0,0 +1,178 @@
+		return throw.New("Object should not be empty")
+	}
+
+	if !m.GetObject().IsSelfScope() {
+		return throw.New("Object reference should be selfScore")
+	}
+
+	if m.GetObject().GetLocal().Pulse() >= currPulse {
+		return throw.New("Object pulse should be less that current pulse")
+	}
+
+	switch m.GetStatus() {
+	case Unknown:
+		return throw.New("Unexepected Unknown status received")
+	case Ready:
+		if err := validateReadyStatus(m, currPulse); err != nil {
+			return err
+		}
+	case Empty:
+		if err := validateEmptyStatus(m, currPulse); err != nil {
+			return err
+		}
+	case Missing:
+	case Inactive:
+		if err := validateZeroPending(m); err != nil {
+			return err
+		}
+
+		if err := validateEmptyState(m); err != nil {
+			return err
+		}
+
+		if err := validateEmptyLatest(m); err != nil {
+			return err
+		}
+
+	}
+
+	return nil
+}
+
+func validateEmptyStatus(m *VStateReport, currPulse PulseNumber) error {
+	if m.GetOrderedPendingCount() != 1 {
+		return throw.New("Should be one ordered pending")
+	}
+
+	if m.GetUnorderedPendingCount() != 0 {
+		return throw.New("Unordered pending count should be 0")
+	}
+
+	if !m.GetUnorderedPendingEarliestPulse().IsUnknown() {
+		return throw.New("Unordered pending earliest pulse should be Unknown")
+	}
+
+	objectPulse := m.GetAsOf()
+	orderedPendingPulse := m.GetOrderedPendingEarliestPulse()
+
+	if orderedPendingPulse < objectPulse || currPulse < orderedPendingPulse {
+		return throw.New("Incorrect pending ordered pulse number")
+	}
+
+	return nil
+}
+
+func validateReadyStatus(m *VStateReport, currPulse PulseNumber) error {
+	if m.GetUnorderedPendingCount() < 0 || m.GetUnorderedPendingCount() >= 127 {
+		return throw.New("UnorderedPendingCount should be in range [0..127)")
+	}
+
+	if m.GetOrderedPendingCount() < 0 || m.GetOrderedPendingCount() >= 127 {
+		return throw.New("OrderedPendingCount should be in range [0..127)")
+	}
+
+	if m.GetUnorderedPendingEarliestPulse() < 0 || m.GetUnorderedPendingEarliestPulse() > currPulse {
+		return throw.New("UnorderedPendingEarliestPulse should be in range (0..currPulse]")
+	}
+
+	if m.GetOrderedPendingEarliestPulse() < 0 || m.GetOrderedPendingEarliestPulse() > currPulse {
+		return throw.New("OrderedPendingEarliestPulse should be in range (0..currPulse]")
+	}
+
+	return nil
+}
+
+func validateZeroPending(m *VStateReport) error {
+	switch {
+	case m.GetOrderedPendingCount() != 0:
+		return throw.New("OrderedPendingCount should be o")
+	case m.GetUnorderedPendingCount() != 0:
+		return throw.New("UnorderedPendingCount should be o")
+	case !m.GetOrderedPendingEarliestPulse().IsUnknown():
+		return throw.New("OrderedPendingEarliestPulse should be Unknown")
+	case !m.GetUnorderedPendingEarliestPulse().IsUnknown():
+		return throw.New("UnorderedPendingEarliestPulse should be Unknown")
+	}
+
+	return nil
+}
+
+func validateEmptyState(m *VStateReport) error {
+	if m.ProvidedContent == nil {
+		return nil
+	}
+
+	switch content := m.ProvidedContent; {
```

Comment:
maybe just always require empty ProvidedContent or create a helper like 
https://play.golang.org/p/fGv_h2MW6UR

Reply:
good point

Reply:
It will not be work because we have a slice in structure

Reply:
it's sad :( and protobuf didn't create IsZero for internal objects >_<